File: jetbrains/src/main/kotlin/deplens/utils/GithubRepoInfoService.kt
package deplens.utils

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.util.concurrency.AppExecutorUtil
import com.intellij.util.proxy.CommonProxy
import deplens.common.RETRY_DELAY_MILLIS
import deplens.common.Result
import deplens.common.ResultWrapper
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.*
import java.net.Proxy
import java.net.URI
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

@Serializable
data class GithubRepoInfo(
    val stars: String,
    val originalStars: Int,
    val updatedDate: String,
    val fetchedAt: Long = System.currentTimeMillis()
)

@Serializable
data class GithubApiResponse(
    val stargazers_count: Int,
    val pushed_at: String,
)

data class RepoKey(val owner: String, val repo: String) {
    override fun toString(): String = "$owner/$repo"
}

object GithubRepoInfoService {

    private val LOG = Logger.getInstance(GithubRepoInfoService::class.java)

    private val requestManager = RequestManager()

    private val json = Json {
        ignoreUnknownKeys = true
    }

    private val httpClient: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .build()
    }

    private val cacheFile: java.io.File by lazy {
        java.io.File(com.intellij.openapi.application.PathManager.getSystemPath(), "deplens/github_repo_cache.json").apply {
            parentFile.mkdirs()
        }
    }

    private val cache = ConcurrentHashMap<String, GithubRepoInfo>()

    init {
        loadCacheFromDisk()
    }

    private fun loadCacheFromDisk() {
        try {
            if (cacheFile.exists()) {
                val content = cacheFile.readText()
                if (content.isNotBlank()) {
                    val map = json.decodeFromString<Map<String, GithubRepoInfo>>(content)
                    val now = System.currentTimeMillis()
                    val threeDaysMillis = 3 * 24 * 60 * 60 * 1000L
                    var hasExpired = false

                    for ((k, v) in map) {
                        if (now - v.fetchedAt > threeDaysMillis) {
                            hasExpired = true
                        } else {
                            cache[k] = v
                        }
                    }

                    if (hasExpired) {
                        saveCacheToDiskAsync()
                    }
                }
            }
        } catch (e: Exception) {
            LOG.warn("Failed to load github repo cache", e)
        }
    }

    private fun saveCacheToDiskAsync() {
        AppExecutorUtil.getAppExecutorService().execute {
            try {
                val content = json.encodeToString(cache.toMap())
                cacheFile.writeText(content)
            } catch (e: Exception) {
                LOG.warn("Failed to save github repo cache", e)
            }
        }
    }

    private val runningRequests = ConcurrentHashMap<String, Call>()

    private val dateFormatter = DateTimeFormatter
        .ofPattern("yyyy-MM-dd")
        .withZone(ZoneId.systemDefault())

    fun getCacheKey(owner: String, repo: String): String {
        return "$owner/$repo"
    }

    fun isFailure(key: String): Boolean {

        return !cache.containsKey(key) && !requestManager.shouldRequest(key)
    }

    fun getRepoKey(path: String): RepoKey? {
        if (!path.startsWith("github.com/")) return null

        val parts = path.split("/")
        if (parts.size < 3) return null

        return RepoKey(parts[1], parts[2])
    }

    fun getRepoInfo(owner: String, repo: String): ResultWrapper<GithubRepoInfo> {
        val key = getCacheKey(owner, repo)
        val info = cache[key]
        if (info != null) {
            val now = System.currentTimeMillis()
            val threeDaysMillis = 3 * 24 * 60 * 60 * 1000L
            if (now - info.fetchedAt > threeDaysMillis) {
                cache.remove(key)
                saveCacheToDiskAsync()
            } else {
                return ResultWrapper(Result.SUCCESS, info)
            }
        }

        return when {
            isFailure(key) -> ResultWrapper(Result.FAILURE, null)
            else -> ResultWrapper(Result.NONE, null)
        }
    }

    fun fetchRepoInfo(owner: String, repo: String, onFinish: (() -> Unit)? = null): Unit {

        val key = getCacheKey(owner, repo)

        val existing = runningRequests[key]
        if (existing != null && !existing.isCanceled()) {
            LOG.debug("Request already running: $key")
            return
        }

        if (!requestManager.shouldRequest(key)) {
            LOG.debug("should not request: $key")
            onFinish?.invoke()
            return
        }

        // TOOD 使用界面配置
        val githubToken = System.getenv("GITHUB_TOKEN") ?: ""

        val request = Request.Builder()
            .url("https://api.github.com/repos/$owner/$repo")
            .header("Accept", "application/vnd.github+json")
            .header("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 13_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36")
            .apply {
                if (githubToken.isNotBlank()) {
                    header("Authorization", "Bearer $githubToken")
                }
            }
            .build()

        val client = httpClient.newBuilder()
            .proxy(resolveProxy())
            .build()

        val call = client.newCall(request)

        runningRequests[key] = call

        try {

            val response = call.execute()

            if (!response.isSuccessful) {
                LOG.warn("GitHub API failed: ${response.code}")
                return
            }

            val body = response.body?.string() ?: return

            val apiResponse = json.decodeFromString<GithubApiResponse>(body)

            val instant = Instant.parse(apiResponse.pushed_at)

            val repoInfo = GithubRepoInfo(
                stars = Formatter.formatGithubStar(apiResponse.stargazers_count),
                originalStars = apiResponse.stargazers_count,
                updatedDate = dateFormatter.format(instant)
            )

            LOG.info("[请求成功] $key, 星数: ${repoInfo.stars}, 更新日期: ${repoInfo.updatedDate}")

            cache[key] = repoInfo
            saveCacheToDiskAsync()

        } catch (e: Exception) {

            LOG.warn("GitHub request error: $key", e)


        } finally {

            if(!cache.containsKey(key)) {
                requestManager.updateFailed(key)

                // 3s后自动重试
                AppExecutorUtil.getAppScheduledExecutorService().schedule(
                    {
                        fetchRepoInfo(owner, repo, onFinish)
                    },
                    RETRY_DELAY_MILLIS,
                    TimeUnit.MILLISECONDS
                )
            }

            runningRequests.remove(key)

        }
    }

    private fun resolveProxy(): Proxy? {
        return try {
            val uri = URI("https://api.github.com")
            val proxies = CommonProxy.getInstance().select(uri)
            proxies.firstOrNull()
        } catch (e: Exception) {
            LOG.warn("Proxy resolve failed", e)
            null
        }
    }

    fun cancelAllRequests() {

        runningRequests.values.forEach {
            if (!it.isCanceled()) {
                it.cancel()
            }
        }

        runningRequests.clear()
    }

    fun clearCache() {
        cache.clear()
        saveCacheToDiskAsync()
    }

    fun shutdown() {
        cancelAllRequests()
        clearCache()
    }
}

File: vscode/src/utils/GithubRepoInfoService.ts
import * as fs from "node:fs/promises";
import * as path from "node:path";
import { fetch } from "undici";
import * as vscode from "vscode";
import { RETRY_DELAY_MILLIS } from "../common/Const";
import { Result, ResultWrapper } from "../common/Result";
import { Formatter } from "./Formatter";
import { Logger } from "./Logger";
import { RequestManager } from "./RequestManager";
import { AbstractCachedRequestService } from "./service/AbstractCachedRequestService";
import { ResultWrapper } from "./service/AbstractCachedRequestService";

export interface GithubRepoInfo {
  stars: string;
  originalStars: number;
  updatedDate: string;
  fetchedAt: number;
}

interface GithubApiResponse {
  stargazers_count: number;
  pushed_at: string;
}

interface NpmPackageInfo {
  stars: string;
  originalStars: number;
  updatedDate: string;
  fetchedAt: number;
}

interface MavenInfo {
  stars: string;
  originalStars: number;
  updatedDate: string;
  fetchedAt: number;
}

interface NpmApiResponse {
  repository?: {
    url?: string;
  };
  time?: {
    modified?: string;
  };
}

interface MavenMetadata {
  // 简化的 Maven 元数据接口
  repositoryUrl?: string;
  lastModified?: string;
}

export class GithubRepoInfoService {
  private static instance: GithubRepoInfoService;
  private cache = new Map<string, GithubRepoInfo>();
  private npmCache = new Map<string, NpmPackageInfo>();
  private mavenCache = new Map<string, MavenInfo>();
  private requestManager = new RequestManager();
  private npmRequestManager = new RequestManager();
  private mavenRequestManager = new RequestManager();
  private logger = new Logger("GithubRepoInfoService");
  private runningRequests = new Map<string, Promise<void>>();
  private npmRunningRequests = new Map<string, Promise<void>>();
  private mavenRunningRequests = new Map<string, Promise<void>>();
  private cacheFile = "";
  private npmCacheFile = "";
  private mavenCacheFile = "";
  private _onDidUpdateRepoInfo = new vscode.EventEmitter<void>();
  public readonly onDidUpdateRepoInfo = this._onDidUpdateRepoInfo.event;

  private constructor() {}

  static getInstance(): GithubRepoInfoService {
    if (!GithubRepoInfoService.instance) {
      GithubRepoInfoService.instance = new GithubRepoInfoService();
    }
    return GithubRepoInfoService.instance;
  }

  async init(context: vscode.ExtensionContext) {
    const storagePath = context.globalStorageUri.fsPath;
    try {
      await fs.mkdir(storagePath, { recursive: true });
    } catch {}
    this.cacheFile = path.join(storagePath, "github_repo_cache.json");
    this.npmCacheFile = path.join(storagePath, "npm_package_cache.json");
    this.mavenCacheFile = path.join(storagePath, "maven_repo_cache.json");
    await this.loadCacheFromDisk();
    await this.loadNpmCacheFromDisk();
    await this.loadMavenCacheFromDisk();
  }

  private async loadCacheFromDisk() {
    try {
      // Check if file exists
      try {
        await fs.access(this.cacheFile);
      } catch {
        return;
      }

      const content = await fs.readFile(this.cacheFile, "utf-8");
      if (content) {
        const map = JSON.parse(content);
        const now = Date.now();
        const threeDaysMillis = 3 * 24 * 60 * 60 * 1000;
        let hasExpired = false;

        for (const key in map) {
          const val = map[key] as GithubRepoInfo;
          if (now - val.fetchedAt > threeDaysMillis) {
            hasExpired = true;
          } else {
            this.cache.set(key, val);
          }
        }

        if (hasExpired) {
          this.saveCacheToDiskAsync();
        }
      }
      } catch (e) {
        this.logger.warn(`Failed to load github repo cache: ${e}`);
      }
  }

  private async loadNpmCacheFromDisk() {
    try {
      try {
        await fs.access(this.npmCacheFile);
      } catch {
        return;
      }

      const content = await fs.readFile(this.npmCacheFile, "utf-8");
      if (content) {
        const map = JSON.parse(content);
        const now = Date.now();
        const threeDaysMillis = 3 * 24 * 60 * 60 * 1000;
        let hasExpired = false;

        for (const key in map) {
          const val = map[key] as NpmPackageInfo;
          if (now - val.fetchedAt > threeDaysMillis) {
            hasExpired = true;
          } else {
            this.npmCache.set(key, val);
          }
        }

        if (hasExpired) {
          this.saveNpmCacheToDiskAsync();
        }
      }
    } catch (e) {
        this.logger.warn(`Failed to load npm package cache: ${e}`);
    }
  }

  private async loadMavenCacheFromDisk() {
    try {
      try {
        await fs.access(this.mavenCacheFile);
      } catch {
        return;
      }

      const content = await fs.readFile(this.mavenCacheFile, "utf-8");
      if (content) {
        const map = JSON.parse(content);
        const now = Date.now();
        const threeDaysMillis = 3 * 24 * 60 * 60 * 1000;
        let hasExpired = false;

        for (const key in map) {
          const val = map[key] as MavenInfo;
          if (now - val.fetchedAt > threeDaysMillis) {
            hasExpired = true;
          } else {
            this.mavenCache.set(key, val);
          }
        }

        if (hasExpired) {
          this.saveMavenCacheToDiskAsync();
        }
      }
    } catch (e) {
        this.logger.warn(`Failed to load maven repo cache: ${e}`);
    }
  }

  private async saveCacheToDiskAsync() {
    try {
      const obj = Object.fromEntries(this.cache);
      await fs.writeFile(this.cacheFile, JSON.stringify(obj));
    } catch (e) {
        this.logger.warn(`Failed to save github repo cache: ${e}`);
    }
  }

  getRepoInfo(owner: string, repo: string): ResultWrapper<GithubRepoInfo> {
    const key = `${owner}/${repo}`;
    const info = this.cache.get(key);
    if (info) {
      const now = Date.now();
      const threeDaysMillis = 3 * 24 * 60 * 60 * 1000;
      if (now - info.fetchedAt > threeDaysMillis) {
        this.cache.delete(key);
        this.saveCacheToDiskAsync();
      } else {
        return { result: Result.SUCCESS, data: info };
      }
    }

    if (this.isFailure(key)) {
      return { result: Result.FAILURE };
    }
    return { result: Result.NONE };
  }

  isFailure(key: string): boolean {
    return !this.cache.has(key) && !this.requestManager.shouldRequest(key);
  }

  async fetchRepoInfo(owner: string, repo: string): Promise<void> {
    const key = `${owner}/${repo}`;
    if (this.runningRequests.has(key)) return;
    if (!this.requestManager.shouldRequest(key)) return;

    const githubToken =
      vscode.workspace.getConfiguration("depLens").get<string>("githubToken") ||
      process.env.GITHUB_TOKEN ||
      "";

    const task = async () => {
      try {
        const resp = await fetch(`https://api.github.com/repos/${owner}/${repo}`, {
          headers: {
            Accept: "application/vnd.github+json",
            "User-Agent": "dep-lens-vscode",
            ...(githubToken ? { Authorization: `Bearer ${githubToken}` } : {}),
          },
        });

        if (!resp.ok) {
          this.logger.warn(`GitHub API failed: ${resp.status}`);
          return;
        }

        const json = (await resp.json()) as GithubApiResponse;
        const dateStr = json.pushed_at;
        const date = new Date(dateStr);
        const formattedDate = date.toISOString().split("T")[0]; // YYYY-MM-DD

        const repoInfo: GithubRepoInfo = {
          stars: Formatter.formatGithubStar(json.stargazers_count),
          originalStars: json.stargazers_count,
          updatedDate: formattedDate,
          fetchedAt: Date.now(),
        };

        this.logger.info(`[Success] ${key}, stars: ${repoInfo.stars}, updated: ${repoInfo.updatedDate}`);

        this.cache.set(key, repoInfo);
        this.saveCacheToDiskAsync();
        this._onDidUpdateRepoInfo.fire();
      } catch (e) {
        this.logger.warn(`GitHub request error: ${key} ${e}`);
      } finally {
        if (!this.cache.has(key)) {
          this.requestManager.updateFailed(key);
          // Retry after delay
          setTimeout(() => {
            this.fetchRepoInfo(owner, repo);
          }, RETRY_DELAY_MILLIS);
        }
        this.runningRequests.delete(key);
      }
    };

    const p = task();
    this.runningRequests.set(key, p);
    return p;
  }

  async clearCache() {
    this.cache.clear();
    await this.saveCacheToDiskAsync();
  }
}
